sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala

```diff
@@ -17,23 +17,44 @@
 package org.apache.spark.sql.execution.datasources
 
 import org.apache.spark.sql.catalyst.analysis.MultiInstanceRelation
-import org.apache.spark.sql.catalyst.expressions.{AttributeMap, AttributeReference}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeMap, AttributeReference}
 import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan, Statistics}
 import org.apache.spark.sql.sources.BaseRelation
 
 /**
  * Used to link a [[BaseRelation]] in to a logical query plan.
+ *
+ * Note that sometimes we need to use `LogicalRelation` to replace an existing leaf node without
+ * changing the output attributes' IDs.  The `expectedOutputAttributes` parameter is used for
+ * this purpose.
  */
-private[sql] case class LogicalRelation(relation: BaseRelation)
-  extends LeafNode
-  with MultiInstanceRelation {
+private[sql] case class LogicalRelation(
+    relation: BaseRelation,
+    expectedOutputAttributes: Seq[Attribute] = Nil)
```

Comment:
Can we use `Option[Seq[Attribute]]`? I think the semantic of having a `Option` at here is more clearer than just using a `Seq` and the meaning of it is self-explanatory. `None` means that `expectedOutputAttributes` is not defined.

If we use `Seq`, `Nil` does not necessarily mean that `expectedOutputAttributes` is not defined. Without other comments, it can mean that we have defined `expectedOutputAttributes` and it is a `Nil`. Readers have to take a look at our comments to understand its meaning.